crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -2,163 +2,231 @@ use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+                error: err.to_string(),
+            }),
+        }
     }
 
-    async fn handle_mpc_proof_init_event(
-        public_parameters: maurer::language::PublicParameters<
-            { maurer::SOUND_PROOFS_REPETITIONS },
-            Lang,
-        >,
-        consensus_adapter: Arc<dyn SubmitToConsensus>,
-        epoch_store: Weak<AuthorityPerEpochStore>,
-        threshold: usize,
+    async fn start_proof_mpc_flow(
+        &self,
+        public_parameters: ProofPublicParameters,
     ) -> anyhow::Result<ProofParty> {
         let batch_size = 1;
-        let party_state: ProofParty =
-            proof::aggregation::asynchronous::Party::new_proof_round_party(
-                public_parameters,
-                PhantomData,
-                threshold as group::PartyID,
-                batch_size,
-                &mut OsRng,
-            )?;
-
-        match party_state.advance(HashMap::new(), &(), &mut OsRng) {
-            Ok(advance_result) => {
-                let AdvanceResult::Advance((message, new_party)) = advance_result else {
-                    return Err(anyhow!("Finalization reached unexpectedly"));
-                };
-                let Some(epoch_store) = epoch_store.upgrade() else {
-                    // TODO: (#259) Handle the case when the epoch switched in the middle of the MPC instance
-                    return Err(anyhow!("Epoch store not found"));
-                };
-                let message_tx = ConsensusTransaction::new_signature_mpc_message(
-                    epoch_store.name,
-                    bcs::to_bytes(&message)?,
-                );
-                consensus_adapter
-                    .submit_to_consensus(&vec![message_tx], &epoch_store)
+        let party: ProofParty = proof::aggregation::asynchronous::Party::new_proof_round_party(
+            public_parameters,
+            PhantomData,
+            self.mpc_threshold_number_of_parties as PartyID,
+            batch_size,
+            &mut OsRng,
+        )?;
+        let Some(epoch_store) = self.epoch_store.upgrade() else {
+            // TODO: (#259) Handle the case when the epoch switched in the middle of the MPC instance
+            return Err(anyhow!("Epoch store not found"));
+        };
+        let Ok(advance_result) = party.advance(HashMap::new(), &(), &mut OsRng) else {
+            // This should never happen, as there should be on chain verification on the initial event input
+            return Err(anyhow!(
+                "Error performing first step for session id: {:?}",
+                self.session_id
+            ));
+        };
+        match advance_result {
+            AdvanceResult::Advance((message, party)) => {
+                let message_tx = self.new_signature_mpc_message(message).ok_or_else(|| {
+                    anyhow!(
+                        "Epoch store switched in the middle of session: {:?}",
+                        self.session_id
+                    )
+                })?;
```

Comment:
Because there's no documentation for `new_signature_mpc_message()`, I can't immediately know that "epoch store switched
in the middle of session" is the only possible reason for it to return `None`, can you justify why that is the case?

Reply:
Added docs